easycv/datasets/detection/data_sources/base.py

```diff
@@ -0,0 +1,171 @@
+            cache_on_the_fly: if set True, will cache in memroy during training
+            parse_fn: parse function to parse source iterator
+            num_processes: number of processes to parse samples
+        """
+        self.CLASSES = classes
+        self.rank, self.world_size = get_dist_info()
+        self.cache_at_init = cache_at_init
+        self.cache_on_the_fly = cache_on_the_fly
+        self.num_processes = num_processes
+
+        if self.cache_at_init and self.cache_on_the_fly:
+            raise ValueError(
+                'Only one of `cache_on_the_fly` and `cache_at_init` can be True!'
+            )
+        source_iter = self.get_source_iterator()
+
+        process_fn = functools.partial(
+            build_sample,
+            parse_fn=parse_fn,
+            classes=self.CLASSES,
+            load_img=cache_at_init == True,
+        )
+        self.samples_list = self.build_samples(
+            source_iter, process_fn=process_fn)
+
+    @abstractmethod
+    def get_source_iterator():
+        """data list iterator, for multi-process read
+        """
+        raise NotImplementedError
+
+    def build_samples(self, iterable, process_fn):
+        samples_list = []
+        with Pool(processes=self.num_processes) as p:
+            with tqdm(total=len(iterable), desc='Scanning images') as pbar:
+                for _, result_dict in enumerate(
+                        p.imap_unordered(process_fn, iterable)):
+                    if result_dict:
+                        samples_list.append(result_dict)
+                    pbar.update()
+
+        return samples_list
+
+    def get_length(self):
+        return len(self.samples_list)
+
+    def __len__(self):
+        return self.get_length()
+
+    def get_ann_info(self, idx):
+        """
+        Get raw annotation info, include bounding boxes, labels and so on.
+        `bboxes` format is as [x1, y1, x2, y2] without normalization.
+        """
+        sample_info = self.samples_list[idx]
+
+        groundtruth_is_crowd = sample_info.get('groundtruth_is_crowd', None)
+        if groundtruth_is_crowd is None:
+            groundtruth_is_crowd = np.zeros_like(sample_info['gt_labels'])
+
+        annotations = {
+            'bboxes': sample_info['gt_bboxes'],
+            'labels': sample_info['gt_labels'],
+            'groundtruth_is_crowd': groundtruth_is_crowd
+        }
+
+        return annotations
+
+    def post_process_fn(self, result_dict):
+        if result_dict.get('img_fields', None) is None:
+            result_dict['img_fields'] = ['img']
+        if result_dict.get('bbox_fields', None) is None:
+            result_dict['bbox_fields'] = ['gt_bboxes']
+
+        return result_dict
+
+    def get_sample(self, idx):
+        result_dict = self.samples_list[idx]
+        load_success = True
+        try:
+            if result_dict.get('img', None) is None:
+                result_dict.update(load_image(result_dict['filename']))
```

Comment:
if cache_at_init == True  image is loaded in init fn, this step should be skipped

Reply:
done!